tracer/src/Datadog.Trace/Util/Http/HttpRequestExtensions.Framework.cs

```diff
@@ -23,30 +25,64 @@ internal static partial class HttpRequestExtensions
             {
                 if (!k.Equals("cookie", System.StringComparison.OrdinalIgnoreCase))
                 {
-                    headersDic.Add(k.ToLowerInvariant(), request.Headers[k]);
+                    var key = k.ToLowerInvariant();
+                    if (!headersDic.ContainsKey(key))
+                    {
+                        headersDic.Add(key, request.Headers[k]);
+                    }
+                    else
+                    {
+                        Log.Warning("Header {key} couldn't be added as argument to the waf", key);
```

Comment:
Isn't it valid for some http headers to appear more than once? Is AppSec ok collecting only the first instance of each?

As an example, the `NameValueCollection` used in (legacy) ASP.NET for request and response headers lets you add multiple values with the same key. If you use `headers[name]` to get a value, it will return them concatenated with commas (which is the standard way to do it in http afaik). Or you can use `headers.GetValues(name)` which returns a `string[]`.

I'm wondering if it makes sense to use `Dictionary<string, List<string>>` for all headers, not only cookies.

Reply:
After testing on IIS and on Net core, all others than cookies end up separated by commas with the same entry in the collection, so I thought this shouldn't happen and that it wasn't worth having the `Dictionary<string, List<string>>` and testing if entry exists etc.

Reply:
So as I use the `headers[name]` here, I indeed get the values comma separated and that's enough to trigger the waf so I thought it's not worth doing it for all, as the bug only happens with cookies on IIS where I get several entries. But as it can technically happen, I can use the list. What is odd is that if I send to IIS twice the same header, it still concatenates them with comma and the GetValues(name) gives me a list with one value, concatenated with commas 🤔, so I'm not sure it's gonna be of much use (?) :
![image](https://user-images.githubusercontent.com/8353486/144024705-a16c3563-5f8a-492a-8c37-033772c6233a.png)

Reply:
That's interesting. Are we sure it's only cookies that get split? (Worth looking into the source?) If so, then I agree, doesn't seem worth having a list!

Reply:
Actually, it's not in System.web so much, because in the code, they end up reading the "ALL_RAW" server variable and IIS gives them comma separated for the same key. So, as we don't have much determinism over that (maybe IIS will behave differently depending on versions and environments) I added arrays for headers instead of a list to avoid converting to list. But basically what I got is:
![image](https://user-images.githubusercontent.com/8353486/144058070-21042efb-f0af-4e5d-836c-00e8aa362cf7.png)

I only handle duplicate keys for cookies, adding to a list if we already found the key, in both core and framework, as someone could run core on IIS..

Reply:
Thanks!